### DISCUSSION TITLE
Many dependent modules have provider version pinned with < 4.0 in terraform-aws-cis-service-catalog

### DISCUSSION THREAD

**Discussion by craigwong** [gruntwork-io/knowledge-base#487] [Help]
Hi there,

We are looking to use [a feature](https://github.com/hashicorp/terraform-provider-aws/pull/24447) in `aws_organizations_account` that requires version 4.15.0 or greater. We have made an update to the `terraform-aws-security` module to use this feature, and in turn have updated `terraform-aws-service-catalog` and `terraform-aws-cis-service-catalog`.

What we've found is that many of the dependent modules have the following provider version constraint:
```hcl
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 3.0.0, < 4.0"
    }
  }
```

which results in a provider version that cannot be queried, because there is no such version `< 4.0 and >= 4.15.0`.

1. Is there a reason for the `< 4.0` constraint such that removing it would result in breakage?
2. What is the recommended way to proceed so that we can use 4.15.0+ provider version (safely)?

Thanks,

---

<ins datetime="2022-06-28T20:51:09Z">
  <p><a href="https://support.gruntwork.io/hc/requests/108854">Tracked in ticket #108854</a></p>
</ins>

**Comment by yorinasub17** [comment]
We're still working through the constraints through our systems. This is mostly due to the backward incompatibility from v3 to v4 and our modules need to be updated to adapt. We are mostly done with the core work in the module catalog (just a few more modules left), and slowly working our way through to the service catalog.

We can't commit to any timelines on when that work would be completed.

However, as a workaround, you can override the version constraints following the strategy listed in https://github.com/gruntwork-io/knowledge-base/discussions/187. AFAIK, this should work out of the box for the modules in `terraform-aws-cis-service-catalog` where there aren't a lot of backward incompatibilities identified, but if there are any that prevent you from deploying them cleanly, then you will have to wait until we complete the work.

**Reply by craigwong** [reply]
Hi @yorinasub17 so I still get this error message:
```hcl
Initializing the backend...

Initializing provider plugins...
- Finding latest version of hashicorp/archive...
- Finding latest version of hashicorp/time...
- Finding hashicorp/aws versions matching ">= 2.58.0, >= 3.0.0, >= 3.13.0, >= 3.29.0, >= 3.41.0, >= 3.66.0, < 4.0.0, >= 4.15.0"...
- Finding latest version of hashicorp/null...
- Installing hashicorp/archive v2.2.0...
- Installed hashicorp/archive v2.2.0 (signed by HashiCorp)
- Installing hashicorp/time v0.7.2...
- Installed hashicorp/time v0.7.2 (signed by HashiCorp)
- Installing hashicorp/null v3.1.1...
- Installed hashicorp/null v3.1.1 (signed by HashiCorp)
╷
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider
│ hashicorp/aws: no available releases match the given constraints >= 2.58.0,
│ >= 3.0.0, >= 3.13.0, >= 3.29.0, >= 3.41.0, >= 3.66.0, < 4.0.0, >= 4.15.0
╵

ERRO[0044] 1 error occurred:
        * exit status 1
```

After I added the `generate` block in the `terragrunt.hcl` in the same directory where I run `terragrunt init`. It calls `terraform-aws-cis-service-catalog` with inputs:
```hcl
generate "provider_version" {
  path      = "provider_version_override.tf"
  if_exists = "overwrite_terragrunt"
  contents  = <<EOF
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 4.15.0"
    }
  }
}
EOF
}
```

Made sure that there are no `.terraform.lock.hcl` files:
```sh
find . -name ".terraform.lock.hcl" -type f #yields no results
```

I can see that the `*_override.tf` gets generated correctly:
```sh
❯ find . -name "provider_version_override.tf" -type f
./.terragrunt-cache/n6Rlq2dStaMegxYPFhI2ofaWeQg/ESyVf0il_W_BBTPCF7EMU2D9LCg/modules/landingzone/account-baseline-root/provider_version_override.tf
❯ cat ./.terragrunt-cache/n6Rlq2dStaMegxYPFhI2ofaWeQg/ESyVf0il_W_BBTPCF7EMU2D9LCg/modules/landingzone/account-baseline-root/provider_version_override.tf
# Generated by Terragrunt. Sig: nIlQXj57tbuaRZEa
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 4.15.0"
    }
  }
}
```

Am I missing a step?

Thanks in advance!

**Reply by yorinasub17** [reply]
Hi Craig, apologies for the toil here. It appears that this is a mistake on my part. Unfortunately, the override file is only able to override the constraints in the top level modules, and not able to override constraints that are defined in the lower level module catalogs. As such, you won't be able to lift the v4 constraint until we lift them from our module catalog.

With that said, I think we are very close to lifting this constraint. Just to confirm, are you looking for this in the `account-baseline-root` module? I'll see if there is anything we can do on our side to prioritize just that module.

**Reply by craigwong** [reply]
Hi @yorinasub17, no problem. Thanks for taking a look. Indeed we are feeding the `create_govcloud` flag to `child_accounts` for the inputs of the `Gruntwork/terraform-aws-cis-service-catalog.git//modules/landingzone/account-baseline-root` module.

That would be great if this could be prioritized! Much appreciated.

**Reply by sms-astanley** [reply]
@yorinasub17 @josh-padnick This Terraform update is a huge improvement for GovCloud compatibility.  We expect it to greatly improve our provisioning time for GovCloud.  I don't want to speak for @craigwong but I'm more than willing to contribute to this feature.

**Reply by josh-padnick** [reply]
@rhoboat Is there any way to leverage @sms-astanley 's team to help out with the remaining AWS Provider v4 updates?

@sms-astanley: We've been working on these for a while and they're quite a pain to do. We even developed a whole new upgrade framework to handle our tests. We're nearly done with these, but not 100% over the finish line yet.

**Reply by craigwong** [reply]
Yep if there's anything I can do to help, let me know. Would be happy to contribute as well.

**Reply by yorinasub17** [reply]
@craigwong We just released https://github.com/gruntwork-io/terraform-aws-cis-service-catalog/releases/tag/v0.39.0, which unlocks the provider for all the `terraform-aws-cis-service-catalog` modules which should resolve this. Thanks for your patience!
